lib/ddtrace/sampling/token_bucket.rb

```diff
@@ -0,0 +1,118 @@
+# TODO: move to a folder outside of 'ddtrace/sampling', as this is technically a generic rate limiter?
+module Datadog
+  module Sampling
+    # TODO: Write class documentation
+    # [Class documentation]
+    class RateLimiter
+      # TODO
+      # @return [Boolean]
+      def allow?(size); end
+
+      # TODO
+      # @return [Float]
+      def effective_rate; end
+    end
+
+    # Implementation of the Token Bucket metering algorithm.
+    #
+    # TODO: Find more canonical link: https://en.wikipedia.org/wiki/Token_bucket
+    #
+    class TokenBucket < RateLimiter
+      attr_reader :rate, :max_tokens
+
+      def initialize(rate, max_tokens = rate)
+        @rate = rate
+        @max_tokens = max_tokens
+
+        @tokens = max_tokens
+        @total_messages = 0
+        @conforming_messages = 0
+        @last_refill = now
+      end
+
+      # Checks if a message of provided +size+
+      # conforms with the current bucket limit.
+      #
+      # If it does, return +true+ and remove +size+
+      # tokens from the bucket.
+      # If it does not, return +false+ without affecting
+      # the tokens form the bucket.
+      #
+      # @return [Boolean] +true+ if message conforms with current bucket limit
+      def allow?(size)
```

Comment:
you could short circuit for `return false if @rate == 0`

Reply:
Done.